README.md
# Introduction

`rethinkdb-job-queue` is a persistent job or task queue backed by [RethinkDB][rethinkdb-url].
It has been built as an alternative to the many queues available on [NPM][npm-search-url].

[![bitHound Overall Score][bithound-overall-image]][bithound-overall-url]
[![bitHound Dependencies][bithound-dep-image]][bithound-dep-url]
[![bitHound Dependencies][bithound-code-image]][bithound-code-url]
[![js-standard-style][js-standard-image]][js-standard-url]
[![NSP Status][nsp-image]][nsp-url]

[![Thinker][thinker-image]][rjq-github-url]

[![NPM][nodei-npm-image]][nodei-npm-url]

Please __Star__ on GitHub / NPM and __Watch__ for updates.

## Features

*   Powered by [RethinkDB][rethinkdb-url]
*   Connect to [multiple databases][queue-connection-url]
*   Create [multiple queues][queue-connection-url]
*   [Distributed worker nodes (PubSub)][queue-pubsub-url]
*   Global [queue events][queue-events-url]
*   Global [job cancellation][job-cancel-url]
*   Global [pause queue][queue-pause-url]
*   Run [concurrent jobs][queue-concurrency-url]
*   Promise based with minimal callbacks
*   [Priority processing of jobs][job-priority-url]
*   [Job progress updates][job-progress-url]
*   [Delayed job start][job-delayed-url]
*   [Find jobs in the queue][find-job-url]
*   [Determine job uniqueness][find-job-url]
*   [Job timeout][job-timeout-url]
*   [Retrying failed jobs][job-retry-url]
*   [Job reanimation][job-reanimation-url]
*   [Job Editing][job-editing-url]
*   Rich job [history log][job-log-url]
*   Over 1400 [integration tests][testing-url]

[queue-connection-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue-Connection
[queue-events-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue-Events
[queue-concurrency-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue-Options#queue-concurrency-option
[queue-pubsub-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue-PubSub
[job-priority-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job-Options#job-priority-option
[job-progress-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job.setProgress
[job-delayed-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Delayed-Job
[find-job-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue.findJob
[job-cancel-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue.process#failed-job-with-cancel
[queue-pause-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue.pause
[job-timeout-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job-Options#job-timeout-option
[job-retry-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job-Retry
[job-reanimation-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job-Editing#job-reanimation
[job-editing-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job-Editing
[job-log-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job.log
[testing-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Testing

## Documentation

For full documentation [please see the wiki][rjq-wiki-url]

See the [Change Log][rjq-changelog-url] for recent changes.

## Project Status

*   This `rethinkdb-job-queue` module is fully functional.
*   There are over 1400 integration tests.
*   Updated to v1.0.0 to support [SemVer](http://semver.org/).

## Quick Start

### Installation

Note: You will need to install [RethinkDB][rethinkdb-url] before you can use `rethinkdb-job-queue`

After installing [RethinkDB][rethinkdb-url] install the job queue using the following command.

```sh
npm install rethinkdb-job-queue --save
```

### Simple Example

```js

const Queue = require('rethinkdb-job-queue')
const qOptions = {
  name: 'Mathematics' // The queue and table name
}
const cxnOptions = {
  db: 'JobQueue', // The name of the database in RethinkDB
}

const q = new Queue(cxnOptions, qOptions)

const job = q.createJob({
  numerator: 123,
  denominator: 456
})

q.process((job, next) => {
    try {
      let result = job.numerator / job.denominator
      // Do something with your result
      return next(result)
    } catch (err) {
      console.error(err)
      return next(err)
    }
})

return q.addJob(job).catch((err) => {
  console.error(err)
})

```

### E-Mail Job Example using [nodemailer][nodemailer-url]

```js

// The following is not related to rethinkdb-job-queue.
// This is the nodemailer configuration
const nodemailer = require('nodemailer')
const transporter = nodemailer.createTransport({
  service: 'Mailgun',
  auth: {
    user: 'postmaster@superheros.com',
    pass: 'your-api-key-here'
  }
})

// Setup e-mail data with unicode symbols
var mailOptions = {
  from: '"Registration" <support@superheros.com>', // Sender address
  subject: 'Registration', // Subject line
  text: 'Click here to complete your registration', // Plaintext body
  html: '<b>Click here to complete your registration</b>' // HTML body
}

// rethinkdb-job-queue configuration
const Queue = require('rethinkdb-job-queue')

// Queue options have defaults and are not required
const qOptions = {
  name: 'RegistrationEmail', // The queue and table name
  masterInterval: 310000, // Database review period in milliseconds
  changeFeed: true, // Enables events from the database table
  concurrency: 100,
  removeFinishedJobs: 2592000000, // true, false, or number of milliseconds
}

// Connection options have defaults and are not required
// You can replace these options with a rethinkdbdash driver object
const cxnOptions = {
  host: 'localhost',
  port: 28015,
  db: 'JobQueue', // The name of the database in RethinkDB
}

// This is the main queue instantiation call
const q = new Queue(cxnOptions, qOptions)

// Customizing the default job options for new jobs
q.jobOptions = {
  priority: 'normal',
  timeout: 300000,
  retryMax: 3, // Four attempts, first then three retries
  retryDelay: 600000 // Time in milliseconds to delay retries
}

const job = q.createJob()
// The createJob method will only create the job locally.
// It will need to be added to the queue.
// You can decorate the job with any data to be saved for processing
job.recipient = 'batman@batcave.com'

q.process((job, next) => {
  // Send email using job.recipient as the destination address
  mailOptions.to = job.recipient
  return transporter.sendMail(mailOptions).then((info) => {
    console.dir(info)
    return next(info)
  }).catch((err) => {
    // This catch is for nodemailer sendMail errors.
    return next(err)
  })
})

return q.addJob(job).then((savedJobs) => {
  // savedJobs is an array of the jobs added with updated properties
}).catch((err) => {
  console.error(err)
})

```

## About the Owner

I, Grant Carthew, am a technologist, trainer, and Dad from Queensland, Australia. I work on code in a number of personal projects and when the need arises I build my own packages.

This project exists because there were no functional job queues built on the RethinkDB database. I wanted an alternative to the other job queues on NPM.

Everything I do in open source is done in my own time and as a contribution to the open source community.

## Contributing

1.  Fork it!
2.  Create your feature branch: `git checkout -b my-new-feature`
3.  Commit your changes: `git commit -am 'Add some feature'`
4.  Push to the branch: `git push origin my-new-feature`
5.  Submit a pull request :D

Please see the [debugging](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Debugging) and [testing](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Testing) documents for more detail.

## Credits

Thanks to the following marvelous packages and people for their hard work:

-   The [RethinkDB][rethinkdb-url] team for the great database.
-   The RethinkDB driver [rethinkdbdash][rethinkdbdash-url] by [Michel][neumino-url]
-   The Promise library [Bluebird][bluebird-url] by [Petka Antonov][petka-url].
-   The [uuid][uuid-url] package.

This list could go on...

## License

MIT

[npm-search-url]: https://npms.io/search?q=job+queue
[rethinkdb-url]: http://www.rethinkdb.com/
[rethinkdbdash-url]: https://github.com/neumino/rethinkdbdash
[neumino-url]: https://github.com/neumino
[rjq-github-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue
[rjq-wiki-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki
[rjq-changelog-url]: https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Change-Log
[thinker-image]: https://cdn.rawgit.com/grantcarthew/node-rethinkdb-job-queue/master/thinkerjoblist.png
[nodemailer-url]: https://www.npmjs.com/package/nodemailer
[bluebird-url]: https://github.com/petkaantonov/bluebird
[petka-url]: https://github.com/petkaantonov
[uuid-url]: https://www.npmjs.com/package/uuid
[bithound-overall-image]: https://www.bithound.io/github/grantcarthew/node-rethinkdb-job-queue/badges/score.svg
[bithound-overall-url]: https://www.bithound.io/github/grantcarthew/node-rethinkdb-job-queue
[bithound-dep-image]: https://www.bithound.io/github/grantcarthew/node-rethinkdb-job-queue/badges/dependencies.svg
[bithound-dep-url]: https://www.bithound.io/github/grantcarthew/node-rethinkdb-job-queue/master/dependencies/npm
[bithound-code-image]: https://www.bithound.io/github/grantcarthew/node-rethinkdb-job-queue/badges/code.svg
[bithound-code-url]: https://www.bithound.io/github/grantcarthew/node-rethinkdb-job-queue
[js-standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg
[js-standard-url]: http://standardjs.com/
[nsp-image]: https://nodesecurity.io/orgs/openjs/projects/3871d340-0ca9-471c-be9a-39df3871262d/badge
[nsp-url]: https://nodesecurity.io/orgs/openjs/projects/3871d340-0ca9-471c-be9a-39df3871262d
[nodei-npm-image]: https://nodei.co/npm/rethinkdb-job-queue.png?downloads=true&downloadRank=true&stars=true
[nodei-npm-url]: https://nodei.co/npm/rethinkdb-job-queue/


CHANGELOG.md
# `rethinkdb-job-queue` Change log

## v1.0.1 / 2016-10-17

*   Fixed `removeFinishedJobs = false` bug.
*   Upgraded  Babel transpiling to [babel-preset-latest](http://babeljs.io/docs/plugins/preset-latest/).

## v1.0.0 / 2016-10-12

*   `Job.update` now adds log entry automatically.
*   Bump to v1.0.0 to support SemVer.

## v0.5.1 / 2016-10-11

*   Dependencies updated.
*   'Job updated' message added.
*   Updated job-add-log module tests.

## v0.5.0 / 2016-10-10

*   Removed `Job.createLog` method.
*   Changed `Job.addLog` signature.
*   Changed `Job.update` signature.
*   `next(error)` stringifies error to `Job.log[].data`.

## v0.4.4 / 2016-10-07

*   Fixed `dateEnable` value when job failed is called.

## v0.4.3 / 2016-10-07

*   Added default log message when data added.
*   Fixed multiple Queue Masters in the same Node.js process.
*   Fixed `queue-summary` query picking up the `State Document`.
*   Fixed `Queue.id` to include database name.

## v0.4.2 / 2016-10-05

*   Added `Queue.findJob` raw option and tests.
*   `db-review` now changes state document state value to `reviewed`.
*   `queue-change` restarts processing on state document `reviewed` state.

## v0.4.1 / 2016-10-03

*   Improved Job constructor to support arrays.
*   Fixed masterInterval option parsing.

## v0.4.0 / 2016-09-30

*   Added `is.log` to the `is` module.
*   Added `Job.setPriority`, `Job.setTimeout`, `Job.setRetryMax`, `Job.setRetryDelay`, `setDateEnable`.
*   Added string, object, or log arguments to `Job.addLog`.
*   Changed `Queue.createJob` API to support data objects, values and options.

## v0.3.0 / 2016-09-28

*   Added global queue pause.
*   Fixed tests for global queue pause.

## v0.2.1 / 2016-09-27

*   Added `Queue.findJob()` method.

## v0.2.0 / 2016-09-27

*   Extended failed job `dateCreated` value in queue-process tests.
*   Added worker stops processing on global `cancelled` event.
*   Refactored `Queue.process`.

## v0.1.2 / 2016-09-20

*   Documentation updated.
*   Tests updated.

## v0.1.1 / 2016-09-19

*   Added the `job-update` module.
*   Added `job.update()` method.
*   Changed `added` job status to `waiting`.

## v0.1.0 / 2016-09-16

*   Added `db-driver` module for connection testing.
*   Changed `Queue()` constructor API to add connection options.
*   Standardized time related options to milliseconds.
*   Added `datetime` module to add and format dates.
*   Updated `is` module to include `dateBefore`, `dateAfter`, `dateBetween`.
*   Removed the `moment` library.

## v0.0.8 / 2016-09-09

*   Replaced `node-uuid` dependency with `uuid` package.
*   Renamed `dateRetry` to `dateEnable`.
*   Updated `is.date()` to remove `moment.isDate()`.
*   Added `Delayed Job` tests to queue-process tests.

## v0.0.7 / 2016-08-23

*   Fixed `next()` calls.
*   Minor refactor.
*   README updated.

## v0.0.6 / 2016-08-23

*   Changed the `next()` function signature.
*   Added `catch()` to `next()` Promise.
*   README updated.

## v0.0.5 / 2016-08-21

*   Changed return values of `Queue.removeJob` to job ids.
*   Removed the `Queue.connection` alias property.
*   Added the `pausing` Queue event.
*   Updated examples to work with the `next()` Promise.
*   Fixed a require path in `queue-add-job`.

## v0.0.4 / 2016-08-18

*   Removed `retry` job priority.
*   Added process timeout extender when progress updated.
*   Added `priorityAsString` to `Job.getCleanCopy`.
*   Fixed `job-options` so it keeps current if only one supplied.
*   Removed `Job.setPayload`.
*   Switched to using `Map()` for queue-process job timeouts.

## v0.0.3 / 2016-08-12

*   Removed `master` Queue options, only using `masterInterval`.
*   `masterInterval` default options changed from 300 sec to 310.
*   Internal parameter rename from `job` or `jobId` to `jobOrId`.
*   Updated README with better examples.

## v0.0.2 / 2016-08-01

*   Removed data from `Queue.createJob()`.
*   Added the `Job.setPayload()` method.

## v0.0.1 / 2016-07-27

*   Initial publish to npm.
*   Changes expected.


src/enums.js
const logger = require('./logger')(module)

const enums = module.exports = {
  priorityFromValue (value) {
    logger(`priorityFromValue: [${value}]`)
    return Object.keys(enums.priority).find(key => enums.priority[key] === value)
  },
  state: {
    docId: '86f6ff5b-0c4e-46ad-9a5f-e90eb19c9b00',
    global: 'global',
    local: 'local'
  },
  priority: {
    lowest: 60,
    low: 50,
    normal: 40,
    medium: 30,
    high: 20,
    highest: 10
  },
  status: {
    // ---------- Queue Status Values ----------
    ready: 'ready',
    processing: 'processing',
    progress: 'progress',
    pausing: 'pausing',
    paused: 'paused',
    resumed: 'resumed',
    removed: 'removed',
    idle: 'idle',
    reset: 'reset',
    error: 'error',
    reviewed: 'reviewed',
    detached: 'detached',
    stopping: 'stopping',
    stopped: 'stopped',
    dropped: 'dropped',
    // ---------- Job Status Values ----------
    created: 'created', // Non-event, initial create job status
    added: 'added', // Event only, not a job status
    waiting: 'waiting', // Non-event, job status only
    active: 'active',
    completed: 'completed',
    cancelled: 'cancelled',
    failed: 'failed',
    terminated: 'terminated',
    log: 'log', // Event only, not a job status
    updated: 'updated' // Event only, not a job status
  },
  options: {
    name: 'rjqJobList',
    host: 'localhost',
    port: 28015,
    db: 'rjqJobQueue',
    masterInterval: 310000, // 5 minutes and 10 seconds
    priority: 'normal',
    timeout: 300000, // 5 minutes
    retryMax: 3,
    retryDelay: 600000, // 10 minutes
    concurrency: 1,
    removeFinishedJobs: 15552000000 // 180 days
  },
  index: {
    indexActiveDateEnable: 'indexActiveDateEnable',
    indexInactivePriorityDateCreated: 'indexInactivePriorityDateCreated',
    indexFinishedDateFinished: 'indexFinishedDateFinished'
  },
  dbResult: {
    deleted: 'deleted',
    errors: 'errors',
    inserted: 'inserted',
    replaced: 'replaced',
    skipped: 'skipped',
    changes: 'changes',
    unchanged: 'unchanged'
  },
  log: {
    information: 'information',
    warning: 'warning',
    error: 'error'
  },
  message: {
    jobAdded: 'Job added to the queue',
    active: 'Job retrieved and active',
    completed: 'Job completed successfully',
    failed: 'Job processing failed',
    cancel: 'Job cancelled by Queue process handler',
    seeLogData: 'See the data attached to this log entry',
    jobUpdated: 'Job updated. Old values in log data',
    jobProgress: 'Job progress updated. Old value in log data',
    jobNotAdded: 'Job not added to the queue',
    jobAlreadyAdded: 'Job is already in the queue',
    jobDataInvalid: 'Job data can not be a function',
    jobInvalid: 'Job object is invalid',
    processTwice: 'Cannot call queue process twice',
    idInvalid: 'The job id is invalid',
    priorityInvalid: 'The job priority value is invalid',
    timeoutInvalid: 'The job timeout value is invalid',
    retryMaxIvalid: 'The job retryMax value is invalid',
    retryDelayIvalid: 'The job retryDelay value is invalid',
    dateEnableIvalid: 'The job dateEnable value is invalid',
    dbError: 'RethinkDB returned an error',
    concurrencyInvalid: 'Invalid concurrency value',
    cancelCallbackInvalid: 'The onCancel callback is not a function',
    globalStateError: 'The global state document change feed is invalid',
    noErrorStack: 'The error has no stack detail',
    noErrorMessage: 'The error has no message'
  }
}
